ding/policy/mbpolicy/mbsac.py

```diff
@@ -0,0 +1,222 @@
+            dist = Independent(Normal(mu, sigma), 1)
+            pred = dist.rsample()
+            action = torch.tanh(pred)
+
+            log_prob = dist.log_prob(
+                pred
+            ) + 2 * (pred + torch.nn.functional.softplus(-2. * pred) - torch.log(torch.tensor(2.))).sum(-1)
+            return action, -self._alpha.detach() * log_prob
+
+        self._actor_fn = actor_fn
+
+        def critic_fn(obss: Tensor, actions: Tensor, model: nn.Module):
+            eval_data = {'obs': obss, 'action': actions}
+            q_values = model.forward(eval_data, mode='compute_critic')['q_value']
+            return q_values
+
+        self._critic_fn = critic_fn
+
+    def _forward_learn(self, data: dict, world_model, envstep) -> Dict[str, Any]:
+        # preprocess data
+        data = default_preprocess_learn(
+            data,
+            use_priority=self._priority,
+            use_priority_IS_weight=self._cfg.priority_IS_weight,
+            ignore_done=self._cfg.learn.ignore_done,
+            use_nstep=False
+        )
+        if self._cuda:
+            data = to_device(data, self._device)
+
+        if len(data['action'].shape) == 1:
+            data['action'] = data['action'].unsqueeze(1)
+
+        self._learn_model.train()
+        self._target_model.train()
+
+        # TODO: use treetensor
+        # rollout length is determined by world_model.rollout_length_scheduler
+        if self._sample_state:
+            # data['reward'], ... are not used
+            obss, actions, rewards, aug_rewards, dones = \
+                world_model.rollout(data['obs'], self._actor_fn, envstep)
+        else:
+            obss, actions, rewards, aug_rewards, dones = \
+                world_model.rollout(data['next_obs'], self._actor_fn, envstep)
+            obss = torch.concat([data['obs'][None, :], obss])
+            actions = torch.concat([data['action'][None, :], actions])
+            rewards = torch.concat([data['reward'][None, :], rewards])
+            aug_rewards = torch.concat([torch.zeros_like(data['reward'])[None, :], aug_rewards])
+            dones = torch.concat([data['done'][None, :], dones])
+
+        # (T+1, B)
+        target_q_values = q_evaluation(obss, actions, partial(self._critic_fn, model=self._target_model))
+        if self._twin_critic:
+            target_q_values = torch.min(target_q_values[0], target_q_values[1]) + aug_rewards
+        else:
+            target_q_values = target_q_values + aug_rewards
+
+        # (T, B)
+        lambda_return = generalized_lambda_returns(target_q_values, rewards, self._gamma, self._lambda, dones[1:])
+
+        # (T, B)
+        # If S_t terminates, we should not consider loss from t+1,...
+        # note: torch.tensor(0).log().exp() = tensor(0.)
+        # example: tensor([0., 0., 1., 0.]).log().cumsum(dim=0).exp() = tensor([1., 1., 0., 0.])
+        weight = (1 - dones[:-1].detach()).log().cumsum(dim=0).exp()
```

Comment:
maybe you can use [this](https://pytorch.org/docs/stable/generated/torch.logsumexp.html?highlight=logsumexp#torch.logsumexp)

Reply:
I find that I can actually use `torch.cumprod` instead of doing `torch.cumsum` in logspace.